### PR TITLE
Semantic Robustness

### PR DESCRIPTION
**Pull request by dxoigmn**
# What does this PR do?

This PR implements a semantic robustness callback for anomalib as described in https://arxiv.org/abs/2405.07969. We make one critical design choice different from the  paper: we supply the model with an augmented image **and mask** for computing metrics. Additionally, anomalib does not implement patching, so all images are square by default. We believe these changes shouldn't significantly alter results since 1) rotating masks is fine so long as we impute pixel mask values as "non-anomalous" or 0 valued.

## Type of change

Please check all relevant options.

- [ ] Improvement (non-breaking)
- [ ] Bug fix (non-breaking)
- [ ] New feature (non-breaking)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

## Testing

Please describe the tests that you ran to verify your changes. Consider listing any relevant details of your test configuration.

- [ ] `pytest`
- [ ] `CUDA_VISIBLE_DEVICES=0 python -m mart experiment=CIFAR10_CNN_Adv trainer=gpu trainer.precision=16` reports 70% (21 sec/epoch).
- [ ] `CUDA_VISIBLE_DEVICES=0,1 python -m mart experiment=CIFAR10_CNN_Adv trainer=ddp trainer.precision=16 trainer.devices=2 model.optimizer.lr=0.2 trainer.max_steps=2925 datamodule.ims_per_batch=256 datamodule.world_size=2` reports 70% (14 sec/epoch).

## Before submitting

- [ ] The title is **self-explanatory** and the description **concisely** explains the PR
- [ ] My **PR does only one thing**, instead of bundling different changes together
- [ ] I list all the **breaking changes** introduced by this pull request
- [ ] I have commented my code
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] New and existing unit tests pass locally with my changes
- [ ] I have run pre-commit hooks with `pre-commit run -a` command without errors

## Did you have fun?

Make sure you had fun coding 🙃

### CODE REVIEW

File: mart/attack/perturber.py [hunk 1]
```diff
@@ -70,6 +72,15 @@ def create_from_tensor(tensor):
...
                 raise NotImplementedError
 
+        # Select input according to index
+        index = []
+        for i in range(max(len(self.shape), len(input.shape))):
+            input_shape = input.shape[i]
+            desired_shape = self.shape[i]
+
+            index.append(slice(0, desired_shape or input_shape))
+        input = input[index]
```
Comment (dxoigmn) [inline mart/attack/perturber.py:82]: This is a terrible implementation but fulfill the spirit of what we need to do to make MART work.

File: mart/callbacks/anomalib.py [hunk 2]
```diff
@@ -0,0 +1,182 @@
...
+            optimizer.step()
+
+        batch["image"] = image_adv.detach()
+
+        return batch
+
+    def attack(self, image, mask, angle, hue, sat):
+        # Rotate image by angle degrees
+        theta = torch.deg2rad(angle)
+        image_rot = rotate_three_pass(image, theta, N=-1, padding_mode="replicate")
```
Comment (mariusarvinte) [inline mart/callbacks/anomalib.py:148]: This may have the in-place modification bug from `rotate_three_pass`, do you need `image.clone()` just to be safe?

Comment (dxoigmn) [reply mart/callbacks/anomalib.py]: i think this is fixed in the latest code.

**Comment by dxoigmn** [comment]
@mzweilin @mariusarvinte: This is ready for review.

File: mart/callbacks/anomalib.py [hunk 3]
```diff
@@ -0,0 +1,425 @@
...
+from __future__ import annotations
+
+import os
+import types
+from copy import deepcopy
+
+import torch
+from lightning.pytorch.callbacks import Callback
+
+from torch_rotation import rotate_three_pass
```
Comment (mariusarvinte) [inline mart/callbacks/anomalib.py:16]: This is a requirement that should be added or somehow specified somewhere.
`pip install torch-rotation==0.1.2`

File: mart/callbacks/anomalib.py [hunk 4]
```diff
@@ -0,0 +1,425 @@
...
+        # FIXME: It would be nice to extract these from trainer.datamodule.test_data.transform
+        image_mean = torch.tensor([0.48145466, 0.4578275, 0.40821073], device=device)
+        image_std = torch.tensor([0.26862954, 0.26130258, 0.27577711], device=device)
+
+        # FIXME: Ideally we would clone the trainer and run fit instead of creating our own optimization loop
+        # Run optimization
+        best_batch = None
+        batch_history = defaultdict(list)
+
+        for step in (pbar := trange(self.steps, desc="Attack", position=1)):
```
Comment (mariusarvinte) [inline mart/callbacks/anomalib.py:104]: This `:=` is a bit of an exotic syntax, doesn't seem to appear anywhere else in the MART repo

Comment (mariusarvinte) [reply mart/callbacks/anomalib.py:104]: There's this: https://github.com/IntelLabs/MART/blob/36609c7f62412a95e401d75f5c6a34f1a2c926d2/mart/callbacks/progress_bar.py#L16 - not sure how helpful it is to re-use

File: mart/callbacks/anomalib.py [hunk 5]
```diff
@@ -0,0 +1,425 @@
...
+            )
+        else:
+            image_adv = rotate(image_adv, angle, mode=mode, padding_mode=padding_mode)
+
+    # Convert image from RGB to HSV space
+    if hue is not None or sat is not None:
+        image_hsv = rgb_to_hsv(image_adv)
+        image_hue, image_sat, image_val = torch.unbind(image_hsv, dim=-3)
+
+    # Additive hue perturbation with STE clipping
```
Comment (mariusarvinte) [inline mart/callbacks/anomalib.py:306]: Comment is inaccurate, we never did STE for `torch.remainder`. I think it's fine though.

File: mart/callbacks/anomalib.py [hunk 6]
```diff
@@ -0,0 +1,425 @@
...
+    if hue is not None:
+        image_hue = image_hue + hue[:, None, None]
+        image_hue = torch.remainder(image_hue, 2 * torch.pi)
+
+    # Additive saturation perturbation with STE clipping
+    if sat is not None:
+        image_sat = image_sat + sat[:, None, None]
+        image_sat = image_sat + (torch.clip(image_sat, 0.0, 1.0) - image_sat).detach()
+
+    # Convert image fro HSV to RGB space
```
Comment (mariusarvinte) [inline mart/callbacks/anomalib.py:316]: "from"

File: mart/callbacks/anomalib.py [hunk 7]
```diff
@@ -0,0 +1,425 @@
...
+    )
+
+    true_positives = mask == 1
+    positive_loss = torch.sum(anomaly_maps * true_positives, dim=(1, 2)) / (
+        torch.sum(true_positives, dim=(1, 2)) + 1e-8
+    )
+
+    loss = negative_loss + positive_loss
+
+    # decrease negatives and increase positives
```
Comment (mariusarvinte) [inline mart/callbacks/anomalib.py:371]: Given that you removed using `maximize=True` in Adam, it may be worth adding an extra comment here that this will be minimized, and the intended effect is exactly opposite than that of the comment.

File: mart/callbacks/anomalib.py (hunk 3 above)
Comment (dxoigmn) [reply mart/callbacks/anomalib.py:16]: Fixed in 2c8c2768f4b76c4b4d94f9330bc90b7783839af6.

File: mart/callbacks/anomalib.py (hunk 4 above)
Comment (dxoigmn) [reply mart/callbacks/anomalib.py:104]: Ah you should learn about the walrus (:= a Python 3.8 feature)! It's super useful in some circumstances.

I think using ProgressBar will be better once this code becomes MART native.

File: mart/callbacks/anomalib.py (hunk 5 above)
Comment (dxoigmn) [reply mart/callbacks/anomalib.py:306]: Fixed in faefb8e. We probably should have used `torch.fmod` though...

File: mart/callbacks/anomalib.py (hunk 6 above)
Comment (dxoigmn) [reply mart/callbacks/anomalib.py]: Fixed in faefb8e53dfb9d7813f81436ccfd41e589f31317.

File: mart/callbacks/anomalib.py (hunk 7 above)
Comment (dxoigmn) [reply mart/callbacks/anomalib.py]: I assume losses are made to be minimized and gains are made to be maximized, but fixed in 85e85bb37907b1a2c3432ee073117786d0a7206a, I think.

**Comment by dxoigmn** [comment]
@mzweilin: Would be nice to get this merged. Really need help on how to package this.

File: mart/callbacks/anomalib.py (hunk 4 above)
Comment (dxoigmn) [reply mart/callbacks/anomalib.py:104]: Good article here about the walrus operator: https://realpython.com/python-walrus-operator/

**Comment by dxoigmn** [comment]
LGTM! Thanks for the README updates!

**Review by mzweilin — APPROVED** [review]
LGTM
